src/ir/match.h

```diff
@@ -0,0 +1,291 @@
+// Returns the top-level modified expression.
+template<class Matcher> Expression* build(Matcher matcher) {
+  Expression* expr = nullptr;
+  matcher.build(&expr);
+  return expr;
+}
+
+// Matchers
+
+// TODO: Make this a template once we have C++20 (needs non-type class template
+// arguments)
+#define LITERAL_MATCHER(matcher, T, typePred, getValue)                        \
+  struct matcher {                                                             \
+    T val;                                                                     \
+    matcher(T val) : val(val) {}                                               \
+                                                                               \
+    bool matches(Expression* expr) const {                                     \
+      auto* c = expr->dynCast<Const>();                                        \
+      return c && typePred && c->value.getValue() == val;                      \
+    }                                                                          \
+  }
+
+LITERAL_MATCHER(i32, int32_t, c->type == Type::i32, geti32);
+LITERAL_MATCHER(i64, int64_t, c->type == Type::i64, geti64);
+LITERAL_MATCHER(ival, int64_t, c->type.isInteger(), getInteger);
+LITERAL_MATCHER(f32, float, c->type == Type::f32, getf32);
+LITERAL_MATCHER(f64, double, c->type == Type::f64, getf64);
+LITERAL_MATCHER(fval, double, c->type.isFloat(), getFloat);
+#undef LITERAL_MATCHER
+
+struct any {
+  Expression** curr;
+
+  any(Expression** curr = nullptr) : curr(curr) {}
+
+  bool matches(Expression* expr) const {
+    if (curr) {
+      *curr = expr;
+    }
+    return true;
+  }
+
+  void build(Expression** outp) const {
+    if (curr) {
+      *outp = *curr;
+    }
+  }
+};
+
+struct constant {
+  Const** curr;
+
+  constant(Const** curr = nullptr) : curr(curr) {}
+
+  bool matches(Expression* expr) const {
+    if (auto* c = expr->dynCast<Const>()) {
+      if (curr) {
+        *curr = c;
+      }
+      return true;
+    }
+    return false;
+  }
+
+  void build(Expression** outp) const {
+    if (curr) {
+      *outp = *curr;
+    }
+  }
+};
+
+template<class ValueMatcher> struct UnaryMatcher {
+  Unary** curr;
+  UnaryOp op;
+  ValueMatcher value;
+
+  UnaryMatcher(Unary** curr, UnaryOp op, ValueMatcher&& value)
+    : curr(curr), op(op), value(std::move(value)) {}
+
+  bool matches(Expression* expr) const {
+    auto* unary = expr->dynCast<Unary>();
+    if (unary && unary->op == op) {
+      if (curr) {
+        *curr = unary;
+      }
+      return value.matches(unary->value);
+    }
+    return false;
+  }
+
+  void build(Expression** outp) const {
+    if (curr) {
+      *outp = *curr;
+    }
+    auto* out = (*outp)->cast<Unary>();
+    out->op = op;
+    value.build(&out->value);
+  }
+};
+
+// TODO: Once we move to C++17, remove these wrapper functions and use the
+// matcher constructors directly like we do with the leaf matchers already.
```

Comment:
What can be done in C++17 exactly? Can you give a code example?

Reply:
IIRC, C++17 introduces template argument deduction for class template arguments. Previously, only function template arguments could be deduced. Since the template argument deduction is critical for the usability of this interface, I had to use wrapper functions where otherwise I could have just used the struct constructors directly.

```
template<class T> struct foo {
  T member;
  foo(T member) : member(member) {}
};

auto myfoo = foo(42); // Illegal before C++17
auto otherfoo = foo<int>(42); // legal

template<T> foo<T> wrapper(T val) {
  return foo<T>(val);
}

auto wrapfoo = wrapper(42); // legal (!)
```